### PR TITLE
Implement api design + GET on /users/<id>

### PR DESCRIPTION
**Pull request by pvaass**

### CODE REVIEW

File: src/ExpandOnline/KlipfolioApi/Object/AbstractObject.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        $this->_data[$name] = $value;
+    }
+
+    /**
+     * @param string $name
+     * @return boolean
+     */
+    public function __isset($name)
+    {
+        return array_key_exists($name, $this->data);
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractObject.php]: _data

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+use ExpandOnline\KlipfolioApi\Request;
+
+/**
+ * Class AbstractCrudObject
+ * @package ExpandOnline\KlipfolioApi\Object
+ */
+abstract class AbstractCrudObject extends AbstractObject
+{
+
+    const FIELD_ID = 'id';
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: docblock

File: src/ExpandOnline/KlipfolioApi/Object/AbstractObject.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    public function getData()
+    {
+        return $this->_data;
+    }
+
+    /**
+     * @return string
+     */
+    public static function className()
+    {
```
Comment (ronoudgenoeg) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractObject.php]: Waarom is dit static?

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+     * @var Client
+     */
+    private $client;
+
+    /**
+     * AbstractCrudObject constructor.
+     * @param null $id
+     * @param Client|null $client
+     */
+    public function __construct($id = null, Client $client = null)
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: This means that applications using the API definetely need to use a factory to create CRUDObjects.

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    public function setId($id)
+    {
+        $this->_data[static::FIELD_ID] = $id;
+    }
+
+
+    /**
+     * @param Client $client The Client instance this object should use to make calls
+     */
+    public function setClient(Client $client)
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: Why?

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    {
+        $this->client = $client;
+    }
+
+    /**
+     * @return Client
+     */
+    public function getClient()
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: protected

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+     *
+     * @param array $params
+     * @return void
+     */
+    public function delete(array $params = array())
+    {
+        $response = $this->getClient()->sendRequest(new Request(
+            $path = $this->getEndpoint() . '/' . $this->_data[static::FIELD_ID],
+            $method = 'GET'
+        )); // TODO: Request voor shit deleten
```
Comment (swordbeta) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: wat

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+        $request = ''; // TODO: Request voor shit posten
+
+        $response = $this->getClient()->sendRequest(new Request(
+            $path = $this->getEndpoint(),
+            $method = 'POST'
+        ));
+
+        $data = $response->getContent();
+
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: newline

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+     * Create function for the object.
+     *
+     * @param array $params Additional parameters to include in the request
+     * @return $this
+     * @throws \Exception
+     */
+    public function create(array $params = array())
+    {
+        if ($this->_data[static::FIELD_ID]) {
+            throw new \Exception("Can't create object that already exists");
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: Custom exception?

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        }
+
+        $request = ''; // TODO: Request voor shit posten
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: ?

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+        $response = $this->getClient()->sendRequest(new Request(
+            $path = $this->getEndpoint(),
+            $method = 'POST'
+        ));
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: Shouldn't `$this->_data?` be included?

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+        $data = $response->getContent();
+
+
+        $id = is_string($data) ? $data : $data[static::FIELD_ID];
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: Which is it?

File: src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+     * Update the object. Function parameters are similar with the create function
+     *
+     * @param array $params Update parameters in assoc
+     * @return $this
+     */
+    public function update(array $params = array())
+    {
+        $response = $this->getClient()->sendRequest(new Request(
+            $path = $this->getEndpoint() . '/' . $this->_data[static::FIELD_ID],
+            $method = 'GET'
```
Comment (kulltc) [inline src/ExpandOnline/KlipfolioApi/Object/AbstractCrudObject.php]: ?

**Pull request merged**